### PR TITLE
ci: split elixir cache to save and restore cache

### LINKED ISSUE #6508: CI - Split the elixir cache composite action to a restore and save action
We have a [composite action that caches our Elixir dependencies](https://github.com/build-trust/ockam/blob/develop/.github/actions/elixir_cache/action.yml), we should split this action to support [saving](https://github.com/actions/cache/blob/main/save/README.md) and [restoring](https://github.com/actions/cache/blob/main/restore/README.md) distinctively, we should `only` save cache `if and only if` the branch is from the default (develop) branch so that we can [cache efficiently](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows), to allow this

- Only use [the `actions/cache` action](https://github.com/build-trust/ockam/blob/353f4be70d0d9455437ce20044e6208e268d306b/.github/actions/elixir_cache/action.yml#L12) (which saves and restore cache) only when the workflow is run as a cron job by adding a [conditional](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution) to ensure that the  [event_name that triggered the workflow](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context) is a [schedule](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule) `${{ github.event_name == 'schedule' }}`
- If worklfow is not run from a scheduler, we should use only [restore cache action](https://github.com/actions/cache/blob/main/restore/README.md)

---

<i>

We love helping new contributors! ❤️
If you have questions or need help as you explore, please join us on [Discord](https://discord.gg/bewvnm6zqS). If you're looking for other issues to contribute to, please checkout our [good first issues](https://github.com/build-trust/ockam/labels/good%20first%20issue).

</i>

### PR DESCRIPTION
**Pull request by ashutosh-rath02**
Fixes #6508 
## Proposed changes
As described in the issue the cache need to be split into save and restore cache

## Checks

- [ ] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [ ] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [ ] There are no Merge commits in this Pull Request. Ockam repo maintains a linear commit history. We merge Pull Requests by rebasing them onto the develop branch. Rebasing to the latest develop branch and force pushing to your Pull Request branch is okay.
- [ ] I have read and accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [ ] I have read and accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/.github/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam/blob/develop/.github/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam](https://github.com/build-trust/ockam) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam/edit/develop/.github/CONTRIBUTORS.csv) file in the github web UI and create a separate Pull Request, this will mark the commit as verified.

### CODE REVIEW

**Comment by mrinalwadhwa** [comment]
@Lucif3r-in Thank you for spending time on this, before we can merge please

- [x] Accept the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding your Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam/blob/develop/.github/CONTRIBUTORS.csv) file in a separate new pull request to the [build-trust/ockam](https://github.com/build-trust/ockam) repository.

The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam/edit/develop/.github/CONTRIBUTORS.csv) file in the github web UI and create a PR, this will mark the commit as verified.

**Comment by ashutosh-rath02** [comment]
Done @mrinalwadhwa 🚀

File: .github/actions/elixir_cache/action.yml [hunk 1]
```diff
@@ -1,19 +1,44 @@
...
-          cache-elixir-${{ github.workflow }}-${{ github.job }}-
+jobs:
+  cache:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check if triggered by schedule
+        run: echo "::set-output name=schedule::${{ github.event_name == 'schedule' }}"
+        id: schedule
+
+      - name: Save Cache
```
Comment (metaclips) [inline .github/actions/elixir_cache/action.yml]: Let's rename this to `Restore and save cache

File: .github/actions/elixir_cache/action.yml [hunk 2]
```diff
@@ -1,19 +1,44 @@
...
+        if: steps.schedule.outputs.schedule == 'true'
```
Comment (metaclips) [inline .github/actions/elixir_cache/action.yml]: we can change this line to `if: ${{ github.event_name == 'schedule' }}`

File: .github/actions/elixir_cache/action.yml [hunk 3]
```diff
@@ -1,19 +1,44 @@
...
-        key: cache-elixir-${{ github.workflow }}-${{ github.job }}-${{ hashFiles('**/mix.lock') }}
-        restore-keys: |
-          cache-elixir-${{ github.workflow }}-${{ github.job }}-
+jobs:
+  cache:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check if triggered by schedule
+        run: echo "::set-output name=schedule::${{ github.event_name == 'schedule' }}"
+        id: schedule
```
Comment (metaclips) [inline .github/actions/elixir_cache/action.yml]: I don't think there's a need for this step, we can call all other steps conditionally as `${{ github.event_name == 'schedule' }}`, please check below review.

File: .github/actions/elixir_cache/action.yml [hunk 4]
```diff
@@ -1,19 +1,44 @@
...
+
+      - name: Save Cache
+        if: steps.schedule.outputs.schedule == 'true'
+        uses: actions/cache@67b839edb68371cc5014f6cea11c9aa77238de78
+        with:
+          path: |
+            **/deps
+            **/_build
+          key: cache-elixir-${{ github.workflow }}-${{ github.job }}-${{ inputs.job_name }}-${{ hashFiles('**/mix.lock') }}
```
Comment (metaclips) [inline .github/actions/elixir_cache/action.yml]: Let's also add a restore cache key
```
restore-keys: |
            cache-elixir-${{ github.workflow }}-${{ github.job }}-${{ inputs.job_name }}-
```

File: .github/actions/elixir_cache/action.yml [hunk 5]
```diff
@@ -1,19 +1,44 @@
...
+      - name: Restore Cache
+        if: steps.schedule.outputs.schedule == 'false'
+        uses: actions/cache@67b839edb68371cc5014f6cea11c9aa77238de78
+        with:
+          path: |
+            **/deps
+            **/_build
+          key: cache-elixir-${{ github.workflow }}-${{ github.job }}-${{ inputs.job_name }}-${{ hashFiles('**/mix.lock') }}
+          restore-keys: |
+            cache-elixir-${{ github.workflow }}-${{ github.job }}-${{ inputs.job_name }}-
```
Comment (metaclips) [inline .github/actions/elixir_cache/action.yml]: For this step, we only want to restore cache `if ${{ github.event_name != 'schedule' }}`, for this, we will only be [using the restore cache action](https://github.com/actions/cache/tree/main/restore#only-restore-cache)

```yaml
  - uses: actions/cache/restore@67b839edb68371cc5014f6cea11c9aa77238de78
    with:
          path: |
            **/deps
            **/_build
          key: cache-elixir-${{ github.workflow }}-${{ github.job }}-${{ inputs.job_name }}-${{ hashFiles('**/mix.lock') }}
          restore-keys: |
            cache-elixir-${{ github.workflow }}-${{ github.job }}-${{ inputs.job_name }}-
```

File: .github/actions/elixir_cache/action.yml [hunk 6]
```diff
@@ -1,19 +1,44 @@
 name: Elixir Cache
 description: Elixir Cache
 
+on:
+  push:
+    branches:
+      - develop
+      #cron data to be provided
+  schedule:
+    - cron: '0 0 * * *'
```
Comment (metaclips) [inline .github/actions/elixir_cache/action.yml]: This is a [composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action), it's an action that's re-usable across multiple workflows, we can set a cron job here as the [`on: event` only triggers at the workflow level](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#on), let's remove this

**Review by metaclips — CHANGES_REQUESTED** [review]
Thanks @Lucif3r-in for taking steps on this PR, few changes left and we should get it merged

**Comment by ashutosh-rath02** [comment]
Hey @metaclips @mrinalwadhwa  sorry for the delay. I have made the requested changes please review😊

**Review by metaclips — APPROVED** [review]

**Review by polvorin — APPROVED** [review]

**Pull request merged**